### PR TITLE
[ready to review] Better nms

### PR DESCRIPTION
**Pull request by ducha-aiki**

### CODE REVIEW

File: kornia/feature/__init__.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
                         harris_response,
                         gftt_response,
                         hessian_response)
-from .nms import NonMaximaSuppression2d, non_maxima_suppression2d
+from .nms import NonMaximaSuppression2d, nms2d, NonMaximaSuppression3d, nms3d
```
Comment (edgarriba) [inline kornia/feature/__init__.py]: i wouldn't do that, add alias if you want but this will break backward compatibility

Comment (edgarriba) [inline kornia/feature/__init__.py]: same for nms3d

File: kornia/feature/scale_space_detector.py [hunk 2]
```diff
@@ -140,13 +151,13 @@ def detect(self, img: torch.Tensor, num_feats: int) -> Tuple[torch.Tensor, torch
             max_coords_best = _scale_index_to_scale(max_coords_best, sigmas_oct)
 
             # Create local affine frames (LAFs)
-            rotmat = angle_to_rotation_matrix(torch.zeros(B, N).to(max_coords_best.device).to(max_coords_best.dtype))
+            rotmat = angle_to_rotation_matrix(torch.zeros(B, N, dtype=dtype, device=dev))
```
Comment (edgarriba) [inline kornia/feature/scale_space_detector.py]: I suggest to replace by : `torch.eye(2).to(dev, dtype).view(1, 1, 2, 2)`

File: kornia/filters/kernels.py [hunk 3]
```diff
@@ -95,6 +95,124 @@ def get_diff_kernel_3x3() -> torch.Tensor:
     ])
 
 
+def get_diff_kernel3d(device=torch.device('cpu'), dtype=torch.float) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/filters/kernels.py]: can you add some documentation ?

File: kornia/filters/kernels.py [hunk 4]
```diff
@@ -95,6 +95,124 @@ def get_diff_kernel_3x3() -> torch.Tensor:
...
+
+                                          [[0.0, 0.0, 0.0],
+                                           [0.0, 0.5, 0.0],
+                                           [0.0, 0.0, 0.0]],
+                                          ],
+                                         ], device=device, dtype=dtype)
+    return kernel.unsqueeze(1)
+
+
+def get_diff_kernel3d_2nd_order(device=torch.device('cpu'), dtype=torch.float) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/filters/kernels.py]: short documentation

File: kornia/filters/sobel.py [hunk 5]
```diff
@@ -65,6 +65,61 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
...
+    Return:
+        torch.Tensor: the spatial gradients of the input feature map.
+
+    Shape:
+        - Input: :math:`(B, C, D, H, W)`
+        - Output: :math:`(B, C, 3, D, H, W)` or :math:`(B, C, 6, D, H, W)`
+
+    Examples:
+        >>> input = torch.rand(1, 3, 4, 4)
+        >>> output = kornia.filters.SpatialGradient()(input)  # 1x3x2x4x4
```
Comment (edgarriba) [inline kornia/filters/sobel.py]: ```output = kornia.filters.SpatialGradient3d()(input)  # 1x3x2x4x4```

File: kornia/filters/sobel.py [hunk 6]
```diff
@@ -65,6 +65,61 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
...
+class SpatialGradient3d(nn.Module):
+    r"""Computes the first  and second order image pyramid derivative in x, y and d using a diff
+    operator.
+
+    Return:
+        torch.Tensor: the spatial gradients of the input feature map.
+
+    Shape:
+        - Input: :math:`(B, C, D, H, W)`
+        - Output: :math:`(B, C, 3, D, H, W)` or :math:`(B, C, 6, D, H, W)`
```
Comment (edgarriba) [inline kornia/filters/sobel.py]: could you specify what's C and D ?

File: kornia/filters/sobel.py [hunk 7]
```diff
@@ -65,6 +65,61 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
...
+        >>> output = kornia.filters.SpatialGradient()(input)  # 1x3x2x4x4
+    """
+
+    def __init__(self,
+                 mode: str = 'diff',
+                 order: int = 1) -> None:
+        super(SpatialGradient3d, self).__init__()
+        self.order: int = order
+        self.mode: str = mode
+        self.kernel = get_spatial_gradient_kernel3d(mode, order)
```
Comment (edgarriba) [inline kornia/filters/sobel.py]: should we force kernel to be without grads ?

File: test/feature/test_nms.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+            [0., 0, 0, 0., 1, 1., 0.],
+            [0., 0, 1.1, 0., 1., 1., 0.],
+            [0., 0, 0, 0., 1., 1., 0.],
+        ]]], device=device).float()
+        nms = kornia.feature.NonMaximaSuppression2d((3, 3)).to(device)
+        scores = nms(inp)
+        assert_allclose(scores, expected, atol=1e-4, rtol=1e-3)
+
+    def test_gradcheck(self, device):
+        k = 0.04
```
Comment (edgarriba) [inline test/feature/test_nms.py]: unused

File: test/feature/test_nms.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+                                   [0., 0., 0., 0., 0.],
+                                   [0., 0., 0., 0., 0.],
+                                   [0., 0., 0., 0., 0.],
+                                   [0., 0., 0., 0., 0.]]]]])
+        nms = kornia.feature.NonMaximaSuppression3d((3, 3, 3)).to(device)
+        scores = nms(inp)
+        assert_allclose(scores, expected, atol=1e-4, rtol=1e-3)
+
+    def test_gradcheck(self, device):
+        k = 0.04
```
Comment (edgarriba) [inline test/feature/test_nms.py]: unused

**Comment by edgarriba** [comment]
don't forget to add to docs

**Comment by ducha-aiki** [comment]
> don't forget to add to docs

Is there a way to add this to the CI tests? I am guilty of this sin often, and not alone )))

**Pull request merged**